File: modules/expense-tracker/src/statement/matcher.py
"""Fuzzy matching for statement line items against Actual Budget transactions.

Scores each uncleared AB transaction candidate by amount proximity,
date proximity, and merchant description token overlap.
Returns top 3 candidate matches above a minimum threshold.
"""


def fuzzy_match(
    stmt_date: str,
    stmt_amount_cents: int,
    stmt_description: str,
    uncleared_txns: list[dict],
) -> list[dict]:
    """Score and rank uncleared AB transactions against a statement line item.

    Args:
        stmt_date: Statement transaction date (YYYY-MM-DD).
        stmt_amount_cents: Statement transaction amount in cents.
        stmt_description: Statement transaction description/merchant.
        uncleared_txns: List of uncleared AB txns, each with keys:
            id, date, amount, payee (or imported_payee).

    Returns:
        Top 3 candidates with score ≥ 50, sorted by score descending.
        Each candidate is { "txn": {...}, "score": int }.
    """
    candidates = []
    stmt_tokens = set(_normalize(stmt_description).split())

    for txn in uncleared_txns:
        score = 0

        txn_amount = txn.get("amount", 0)
        if txn_amount == stmt_amount_cents:
            score += 80
        elif abs(txn_amount - stmt_amount_cents) <= 20:
            score += 50

        txn_date = txn.get("date", "")
        date_diff = _date_diff(stmt_date, txn_date)
        if date_diff == 0:
            score += 30
        elif date_diff <= 2:
            score += 15

        txn_payee = txn.get("payee") or txn.get("imported_payee") or ""
        txn_tokens = set(_normalize(txn_payee).split())
        if _jaccard(stmt_tokens, txn_tokens) > 0.5:
            score += 20

        if score < 50:
            continue

        candidates.append({"txn": txn, "score": score})

    candidates.sort(key=lambda c: c["score"], reverse=True)
    return candidates[:3]


def _normalize(s: str) -> str:
    return s.lower().strip()


def _date_diff(d1: str, d2: str) -> int:
    if not d1 or not d2:
        return 99
    try:
        from datetime import date
        parts1 = [int(x) for x in d1.split("-")]
        parts2 = [int(x) for x in d2.split("-")]
        return abs((date(*parts1) - date(*parts2)).days)
    except (ValueError, TypeError):
        return 99


def _jaccard(a: set, b: set) -> float:
    if not a or not b:
        return 0.0
    return len(a & b) / len(a | b)

File: modules/expense-tracker/src/statement/matcher.js
/**
 * Fuzzy matching for statement line items against Actual Budget transactions.
 * Ported 1:1 from src/statement/matcher.py
 *
 * Scores each uncleared AB transaction candidate by amount proximity,
 * date proximity, and merchant description token overlap.
 * Returns top 3 candidate matches above a minimum threshold of 50.
 */

/**
 * Score and rank uncleared AB transactions against a statement line item.
 *
 * @param {string} stmtDate - Statement transaction date (YYYY-MM-DD)
 * @param {number} stmtAmountCents - Statement transaction amount in cents
 * @param {string} stmtDescription - Statement transaction description/merchant
 * @param {Array<object>} unclearedTxns - List of uncleared AB txns, each with:
 *   id, date, amount, payee (or imported_payee)
 * @returns {Array<{ txn: object, score: number }>} Top 3 candidates with score >= 50
 */
export function fuzzyMatch(
    stmtDate,
    stmtAmountCents,
    stmtDescription,
    unclearedTxns,
) {
    const candidates = [];
    const stmtTokens = new Set(_normalize(stmtDescription).split(/\s+/));

    for (const txn of unclearedTxns) {
        let score = 0;

        const txnAmount = txn.amount ?? 0;
        if (txnAmount === stmtAmountCents) {
            score += 80;
        } else if (Math.abs(txnAmount - stmtAmountCents) <= 20) {
            score += 50;
        }

        const txnDate = txn.date || "";
        const dateDiff = _dateDiff(stmtDate, txnDate);
        if (dateDiff === 0) {
            score += 30;
        } else if (dateDiff <= 2) {
            score += 15;
        }

        const txnPayee = txn.payee || txn.imported_payee || "";
        const txnTokens = new Set(_normalize(txnPayee).split(/\s+/));
        if (_jaccard(stmtTokens, txnTokens) > 0.5) {
            score += 20;
        }

        if (score < 50) continue;

        candidates.push({ txn, score });
    }

    candidates.sort((a, b) => b.score - a.score);
    return candidates.slice(0, 3);
}

export function _normalize(s) {
    return s.toLowerCase().trim();
}

export function _dateDiff(d1, d2) {
    if (!d1 || !d2) return 99;
    try {
        const parts1 = d1.split("-").map(Number);
        const parts2 = d2.split("-").map(Number);
        if (parts1.length < 3 || parts2.length < 3) return 99;
        if (parts1.some(isNaN) || parts2.some(isNaN)) return 99;
        const date1 = new Date(parts1[0], parts1[1] - 1, parts1[2]);
        const date2 = new Date(parts2[0], parts2[1] - 1, parts2[2]);
        if (isNaN(date1.getTime()) || isNaN(date2.getTime())) return 99;
        return Math.abs(Math.round((date2 - date1) / (1000 * 60 * 60 * 24)));
    } catch {
        return 99;
    }
}

export function _jaccard(a, b) {
    if (!a.size || !b.size) return 0.0;
    const intersection = new Set([...a].filter((x) => b.has(x)));
    return intersection.size / (a.size + b.size - intersection.size);
}
